### ISSUE TITLE
Navigation issue after upgrading from Vaadin 14.2.0 to 14.2.1 (com.vaadin.flow.router.PreserveOnRefresh)

### ISSUE THREAD

**Issue by TGit911iger** [vaadin/flow#8605]
I am using Vaadin 14.2.0 with @com.vaadin.flow.router.PreserveOnRefresh.

Now I am navigation to one of my views and the framework calls

    @Override
    public void setParameter(final com.vaadin.flow.router.BeforeEvent newEvent,
        @com.vaadin.flow.router.WildcardParameter final java.lang.String newParameter) {

Using 14.2.0 works fine the function setParameter ist called once a time during navigation.

Also browser refresh is working very well.

Now I am upgrading to 14.2.1 - the issue occurs. The function setParameter is called two times during navigation and the gui flickers. Why it is called two times?

If I uncommend PreserveOnRefresh all is fine but browser refresh does not work anymore.

The same problem occurs using Vaadin 16 and 17.

Only 14.2.0 is working for me.

I think that the problem came with

flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java

https://github.com/vaadin/flow/blob/master/flow-server/src/main/java/com/vaadin/flow/router/internal/AbstractNavigationStateRenderer.java

Line 790 seems to be the problem: "prevUi.close();"

            // Transfer all remaining UI child elements (typically dialogs
            // and notifications) to the new UI
            maybePrevUI.ifPresent(prevUi -> {
                ui.getInternals().moveElementsFrom(prevUi);
                prevUi.close();
            });

My ui is de.thomas.MyClass.

A lot of RouterLinks are directing to de.thomas.MyClass (UI 1, UI 2, UI 3).

UI 1 with URL parameter "my-parameter=1". UI 2 with URL parameter "my-parameter=2". UI 3 with URL parameter "my-parameter=3".

I am showing UI 1 for the first time - all is OK.

I put a break point in

    public void setParameter(final com.vaadin.flow.router.BeforeEvent newEvent,
        @com.vaadin.flow.router.WildcardParameter final java.lang.String newParameter) {

There I see my expected paramter "my-parameter=1".

Now I am showing UI 2 (following the RouterLink). An then it becomes very strange.

It stops 2 times in the method "setParameter" (url parameters are ok "my-parameter=2"). The ui is shown 2 times and the ui flickers. Also I put a break point in line 790 of class "AbstractNavigationStateRenderer". There it stops 2 times too.

I think line 790 is the problem. 790 came with Vaadin 14.2.1 - 14.2.0 ist working.

Versions:
    - Greater than 14.2.0
    - All 11.0.7
    - macOS 10.15.5
    - All browsers
    - WildFly 19
    - Eclipse 2020-06

Now I took your Business App starter and changed the code to reproduce the issue.

Here are my changes in detail:

```
@com.vaadin.flow.router.PreserveOnRefresh
public class MainLayout extends FlexBoxLayout implements RouterLayout, AfterNavigationObserver {
    /**
     * Initialise the navigation items.
     */
    private void initNaviItems() {
        final NaviMenu menu = this.naviDrawer.getMenu();
        menu.addNaviItem(VaadinIcon.HOME, "Home", Home.class);
        menu.addNaviItem(VaadinIcon.INSTITUTION, "Accounts", Accounts.class);
        menu.addNaviItem(VaadinIcon.CREDIT_CARD, "Payments", Payments.class);
        menu.addNaviItem(VaadinIcon.CHART, "Statistics", Statistics.class);

        final NaviItem personnel = menu.addNaviItem(VaadinIcon.USERS, "Personnel", null);
        menu.addNaviItem(personnel, "Accountants A", Accountants.class, "My parameter 1");
        menu.addNaviItem(personnel, "Accountants B", Accountants.class, "My parameter 2");
        menu.addNaviItem(personnel, "Accountants C", Accountants.class, "My parameter 3");
        menu.addNaviItem(personnel, "Accountants D", Accountants.class, "My parameter 4");
        menu.addNaviItem(personnel, "Accountants E", Accountants.class, "My parameter 5");
        menu.addNaviItem(personnel, "Accountants F", Accountants.class, "My parameter 6");
        menu.addNaviItem(personnel, "Accountants G", Accountants.class, "My parameter 7");
        menu.addNaviItem(personnel, "Accountants H", Accountants.class, "My parameter 8");
        menu.addNaviItem(personnel, "Accountants I", Accountants.class, "My parameter 9"); 
        menu.addNaviItem(personnel, "Managers", Managers.class);
    }
    
    
public class Accountants extends ViewFrame implements com.vaadin.flow.router.HasUrlParameter<java.lang.String> {
    /**
     * setParameter
     */
    @Override
    public void setParameter(final com.vaadin.flow.router.BeforeEvent newEvent,
            @com.vaadin.flow.router.WildcardParameter final java.lang.String newParameter) {

    }
    
    
public class NaviMenu extends Nav {    
    public NaviItem addNaviItem(final NaviItem parent, final String text,
            final Class<? extends Component> navigationTarget, final java.lang.String myParameter) {
        final NaviItem item = new NaviItem(text, navigationTarget, myParameter);
        this.addNaviItem(parent, item);
        return item;
    }
    
    
public class NaviItem extends ListItem {
    public NaviItem(final String text, final Class<? extends Component> navigationTarget,
            final java.lang.String myParameter) {
        this.setClassName(this.CLASS_NAME);
        this.setLevel(0);

        this.text = text;
        this.navigationTarget = navigationTarget;

        if (navigationTarget != null) {
            final java.util.HashMap<java.lang.String, java.lang.String> queryParameter = new java.util.HashMap<java.lang.String, java.lang.String>();
            queryParameter.put("my_own_test_parameter", myParameter);

            final com.vaadin.flow.router.QueryParameters qp = com.vaadin.flow.router.QueryParameters
                    .simple(queryParameter);

            final RouterLink routerLink = new RouterLink(null, navigationTarget);
            routerLink.add(new Span(text));
            routerLink.setClassName(this.CLASS_NAME + "__link");
            routerLink.setHighlightCondition(HighlightConditions.sameLocation());
            routerLink.setQueryParameters(qp);

            this.link = routerLink;
        } else {
            final Div div = new Div(new Span(text));
            div.addClickListener(e -> this.expandCollapse.click());
            div.setClassName(this.CLASS_NAME + "__link");
            this.link = div;
        }

        this.expandCollapse = UIUtils.createButton(VaadinIcon.CARET_UP, ButtonVariant.LUMO_SMALL,
                ButtonVariant.LUMO_TERTIARY);
        this.expandCollapse.addClickListener(event -> this.setSubItemsVisible(!this.subItemsVisible));
        this.expandCollapse.setVisible(false);

        this.subItems = new ArrayList<>();
        this.subItemsVisible = true;
        this.updateAriaLabel();

        this.add(this.link, this.expandCollapse);
    }
```

I attached my sample project.

Just klick on "Accountants A", "Accountants A", "Accountants C" - UI is flickering and the setParameter(...) is called twice.

Thank you very much for your help,
Thomas
[business-app-starter.zip](https://github.com/vaadin/flow/files/4824484/business-app-starter.zip)

**Comment by pleku** [comment]
Thanks for the issue. It would look like this is a regression caused by most likely https://github.com/vaadin/flow/commit/79ff40505f99d22d93c287d8f7559f379c810720

**Comment by denis-anisimov** [comment]
Thank you for problem analysis but the fix may not be verified without exact steps to reproduce.
Could you please provide the steps to reproduce ?

**Comment by denis-anisimov** [comment]
The project contains a lot of extra information which has no any relation to the ticket.
I will have to waste time to clean up it.

I'm not able to reproduce the described behavior with easy view:

```
@PreserveOnRefresh
@Route("com.vaadin.flow.uitest.ui.PreserveOnRefreshHasParamView")
public class PreserveOnRefreshHasParamView extends Div
        implements HasUrlParameter<String> {

    public PreserveOnRefreshHasParamView() {
        setText("Loaded");
    }

    @Override
    public void setParameter(BeforeEvent event, String parameter) {
        System.out.println("set param is called");
    }

}
```

**Comment by denis-anisimov** [comment]
With the attached project I see the described behavior.
Now everything which is not related to the issue needs to be removed ( almost everything).

**Comment by TGit911iger** [comment]
Hello!

Here is a smaller test project to reproduce the error. I hope this will help. Just click on the items from A to F. You will see my output in the console. There are two screenshots attached. One with PreserveOnRefresh and one without. 

Thank you very much for your help, Thomas
[business-app-starter-8605.zip](https://github.com/vaadin/flow/files/4831143/business-app-starter-8605.zip)
<img width="1071" alt="With-PreserveOnRefresh" src="https://user-images.githubusercontent.com/9194576/85719608-de39f280-b6ef-11ea-8633-204f17b6ea15.png">
<img width="1071" alt="Without-PreserveOnRefresh" src="https://user-images.githubusercontent.com/9194576/85719628-e2fea680-b6ef-11ea-8786-ab95a9b45ff7.png">

**Comment by TGit911iger** [comment]
Hello!

Thank you very much for your help but in 14.2.0 it worked. What is about apps using the same ui in multiple instances? We have got multiple instances in one tree and now our app does not work anymore pressing browser refresh.

Kind regards, Thomas

**Comment by joheriks** [comment]
@TGit911iger Verified that with the fix in latest Flow 3.2 snapshot build the `setParameter` method is called only once when `@PreserveOnRefresh` is active in the attached example [app to reproduce](https://github.com/vaadin/flow/files/4831143/business-app-starter-8605.zip). Did you find a new issue after this fix that warrant re-opening the ticket (not sure what using the same ui in multiple instances meant)?

**Comment by denis-anisimov** [comment]
@TGit911iger , sorry , I don't understand your last comment.
The issue is fixed but it's not yet released.
Please wait for release and check.
You may use SNAPSHOT for testing as well.

If you have another issue then please create another ticket and provide the steps to reproduce.

**Comment by TGit911iger** [comment]
Hi!

Oh, I am sorry. Please ignore my comment. I was too fast with my answer. I will wait until official release.

Thank you very much for your help.

Thomas

**Comment by TGit911iger** [comment]
Hello!

Now the first problem as I described is solved - thank you very much. Navigation in tree is working.

But if you press browser refresh in the new Vaadin 14.2.3 it is showing a JavaScript error "Assertion error: No child node found with id 4". In 14.2.0 this error does not occur.

Maybe a bug.

Thank you very much, Thomas
